README.md
# LombScargle.jl

| **Documentation**                       | **Build Status**                    | **Code Coverage**               |
|:---------------------------------------:|:-----------------------------------:|:-------------------------------:|
| [![][docs-stable-img]][docs-stable-url] | [![Build Status][gha-img]][gha-url] | [![][coveral-img]][coveral-url] |
| [![][docs-latest-img]][docs-latest-url] |                                     | [![][codecov-img]][codecov-url] |

Introduction
------------

`LombScargle.jl` is a [Julia](http://julialang.org/) package for a fast
multi-threaded estimation of
the [frequency spectrum](https://en.wikipedia.org/wiki/Frequency_spectrum) of a
periodic signal
with
[the Lomb–Scargle periodogram](https://en.wikipedia.org/wiki/The_Lomb–Scargle_periodogram).

Another Julia package that provides tools to perform spectral analysis of
signals is [`DSP.jl`](https://github.com/JuliaDSP/DSP.jl), but its methods
require that the signal has been sampled at equally spaced times.  Instead, the
Lomb–Scargle periodogram enables you to analyze unevenly sampled data as well,
which is a fairly common case in astronomy, a field where this periodogram is
widely used.

The algorithms used in this package are reported in the following papers:

* Press, W. H., Rybicki, G. B. 1989, ApJ, 338, 277 (URL:
  http://dx.doi.org/10.1086/167197, Bibcode:
  http://adsabs.harvard.edu/abs/1989ApJ...338..277P)
* Townsend, R. H. D. 2010, ApJS, 191, 247 (URL:
  http://dx.doi.org/10.1088/0067-0049/191/2/247, Bibcode:
  http://adsabs.harvard.edu/abs/2010ApJS..191..247T)
* Zechmeister, M., Kürster, M. 2009, A&A, 496, 577 (URL:
  http://dx.doi.org/10.1051/0004-6361:200811296, Bibcode:
  http://adsabs.harvard.edu/abs/2009A%26A...496..577Z)

The package provides facilities to:

* compute the periodogram using different methods (with different speeds) and
  different normalizations.  This is one of the fastest implementations of these
  methods available as free software.  If Julia is run with more than
  one
  [thread](http://docs.julialang.org/en/stable/manual/parallel-computing/#multi-threading-experimental),
  computation is automatically multi-threaded, further speeding up calculations;
* access the frequency and period grid of the resulting periodogram, together
  with the power spectrum;
* find the maximum power in the periodogram and the frequency and period
  corresponding to the peak.  All these queries can be restricted to a specified
  region, in order to search a local maximum, instead of the global one;
* calculate the probability that a peak arises from noise only (false-alarm
  probability) using analytic formulas, in order to assess the significance of
  the peak;
* perform bootstrap resamplings in order to compute the false-alarm probability
  with a statistical method;
* determine the best-fitting Lomb–Scargle model for the given data set at the
  given frequency.

All these features are thoroughly described in the full documentation, see
below.  Here we only give basic information.

### Documentation

The complete manual of `LombScargle.jl` is available
[here](https://juliaastro.github.io/LombScargle.jl/stable/).  It has detailed explanation of all
functions provided by the package and more examples than what you will find
here, also with some plots.

Installation
------------

The latest version of `LombScargle.jl` is available for Julia 1.0 and later
versions, and can be installed with [Julia built-in package
manager](https://julialang.github.io/Pkg.jl/stable/).  In a Julia session, after
entering the package manager mode with `]`, run the command

```julia
pkg> add LombScargle
```

Older versions are also available for Julia 0.4-0.7.

Usage
-----

After installing the package, you can start using it with

```julia
julia> using LombScargle
```

The module defines a new `LombScargle.Periodogram` data type, which, however, is
not exported because you will most probably not need to directly manipulate such
objects.  This data type holds both the frequency and the power vectors of the
periodogram.

The main function provided by the package is `lombscargle`:

```julia
lombscargle(times, signal[, errors])
```

which returns a `LombScargle.Periodogram`.  The only mandatory arguments are:

* `times`: the vector of observation times
* `signal`: the vector of observations associated with `times`

All these vectors must have the same length.  The only optional argument is:

* `errors`: the uncertainties associated to each `signal` point.  This vector
  must have the same length as `times` and `signal`.

Besides the two arguments introduced above, `lombscargle` has a number of other
optional keywords in order to choose the right algorithm to use and tweak the
periodogram.  For the description of all these arguments see the complete
manual.

If the signal has uncertainties, the `signal` vector can also be a vector of
`Measurement` objects (from
[`Measurements.jl`](https://github.com/JuliaPhysics/Measurements.jl) package), in
which case you need not to pass a separate `errors` vector for the uncertainties
of the signal.  You can create arrays of `Measurement` objects with the
`measurement` function, see `Measurements.jl` manual at
https://juliaphysics.github.io/Measurements.jl/latest/ for more details.

With the `LombScargle.plan` function you can pre-plan a periodogram and save
time and memory for the actual computation of the periodogram.  See the
[manual](https://juliaastro.github.io/LombScargle.jl/stable/#Planning-the-Periodogram-1)
for details.

Examples
--------

Here is an example of a noisy periodic signal (`sin(π*t) + 1.5*cos(2π*t)`)
sampled at unevenly spaced times.

```julia
julia> using LombScargle

julia> ntimes = 1001
1001

# Observation times
julia> t = range(0.01, stop=10pi, length=ntimes)
0.01:0.03140592653589793:31.41592653589793

# Randomize times
julia> t += step(t)*rand(ntimes);

# The signal
julia> s = sinpi.(t) .+ 1.5 .* cospi.(2t) .+ rand(ntimes);

# Pre-plan the periodogram (see the documentation)
julia> plan = LombScargle.plan(t, s);

# Compute the periodogram
julia> pgram = lombscargle(plan)
```

You can plot the result, for example with
[`Plots`](https://github.com/tbreloff/Plots.jl) package.  Use `freqpower`
function to get the frequency grid and the power of the periodogram as a
2-tuple.

```julia
using Plots
plot(freqpower(pgram)...)
```

### Signal with Uncertainties

The generalised Lomb–Scargle periodogram (used when the `fit_mean` optional
keyword is `true`) is able to handle a signal with uncertainties, and they will
be used as weights in the algorithm.  The uncertainties can be passed either as
the third optional argument `errors` to `lombscargle` or by providing this
function with a `signal` vector of type `Measurement` (from
[`Measurements.jl`](https://github.com/JuliaPhysics/Measurements.jl) package).

```julia
using Measurements, Plots
ntimes = 1001
t = range(0.01, stop=10pi, length=ntimes)
s = sinpi.(2t)
errors = rand(0.1:1e-3:4.0, ntimes)
plot(freqpower(lombscargle(t, s, errors, maximum_frequency=1.5))...)
plot(freqpower(lombscargle(t, measurement(s, errors), maximum_frequency=1.5))...)
```

Performance
-----------

A pre-planned periodogram in `LombScargle.jl` computed in single thread mode
with the fast method is more than 2 times faster than the implementation of the
same algorithm provided by Astropy, and more than 4 times faster if 4 FFTW
threads are used (on machines with at least 4 physical CPUs).

The following plot shows a comparison between the times needed to compute a
periodogram for a signal with N datapoints using `LombScargle.jl`, with 1 or 4
FFTW threads (with `flags = FFTW.MEASURE` for better performance), and the
single-threaded Astropy implementation.  (Julia version: 1.6.0; `LombScargle.jl`
version: 1.0.0; Python version: 3.8.6; Astropy version: 4.1.  CPU: Intel(R)
Core(TM) i7-4870HQ CPU @ 2.50GHz.)

![benchmarks](https://raw.githubusercontent.com/JuliaAstro/LombScargle.jl/master/perf/benchmarks.svg)

Note that this comparison is unfair, as Astropy doesn’t support pre-planning a
periodogram nor multi-threading, and it pads vectors for FFT to a length which
is a power of 2, while by default `LombScargle.jl` uses length which are
multiples of 2, 3, 5, 7.  A non-planned periodogram in single thread mode in
`LombScargle.jl` is still twice as fast as Astropy.

Development
-----------

The package is developed at https://github.com/JuliaAstro/LombScargle.jl.  There
you can submit bug reports, make suggestions, and propose pull requests.

### History

The ChangeLog of the package is available in
[NEWS.md](https://github.com/JuliaAstro/LombScargle.jl/blob/master/NEWS.md) file
in top directory.

License
-------

The `LombScargle.jl` package is licensed under the BSD 3-clause "New" or
"Revised" License.  The original author is Mosè Giordano.

### Acknowledgemets

This package adapts the implementation in Astropy of the the fast Lomb–Scargle
method by Press & Rybicki (1989).  We claim no endorsement nor promotion by the
Astropy Team.



[docs-latest-img]: https://img.shields.io/badge/docs-latest-blue.svg
[docs-latest-url]: https://juliaastro.github.io/LombScargle.jl/latest/

[docs-stable-img]: https://img.shields.io/badge/docs-stable-blue.svg
[docs-stable-url]: https://juliaastro.github.io/LombScargle.jl/stable/

[gha-img]: https://github.com/JuliaAstro/LombScargle.jl/workflows/CI/badge.svg
[gha-url]: https://github.com/JuliaAstro/LombScargle.jl/actions?query=workflow%3ACI

[coveral-img]: https://coveralls.io/repos/github/JuliaAstro/LombScargle.jl/badge.svg?branch=master
[coveral-url]: https://coveralls.io/github/JuliaAstro/LombScargle.jl?branch=master

[codecov-img]: https://codecov.io/gh/JuliaAstro/LombScargle.jl/branch/master/graph/badge.svg
[codecov-url]: https://codecov.io/gh/JuliaAstro/LombScargle.jl


NEWS.md
# History of LombScargle.jl

## v1.0.3 (2021-11-23)

### Bug Fixes

* Removed some redundant internal methods.
* Fixed a typo in PSD normalisation formula in documentation
  ([#28](https://github.com/JuliaAstro/LombScargle.jl/issues/28),
  [#29](https://github.com/JuliaAstro/LombScargle.jl/pull/29)).

## v1.0.2 (2021-08-19)

### Bug Fixes

* Make sure `bootstrap([::AbstractRNG,] ::Integer, ::PeriodogramPlan)` methods
  don't modify the periodogram of the input plan.

## v1.0.1 (2021-08-17)

### Bug Fixes

* Restore `bootstrap(::Integer, ::PeriodogramPlan)` method that was accidentally
  removed in v0.5.1.

## v1.0.0 (2020-11-22)

### New Features

* Previously, when using the fast method the vector of the signal was padded to
  a length which is a power of 2, but now you can choose the factors of the
  length of the padded vector with the new keyword argument for
  `LombScargle.lombscargle` and `LombScargle.plan` functions:
  `padding_factors::Union{NTuple{N,<:Integer} where {N},Vector{<:Integer}}`.
  This defaults to `[2,3,5,7]`, which is the optimal choice for FFTW and allows
  for smaller vectors compared to powers of 2.  To reproduce the same results as
  with the previous default setting you need to use `padding_factors=[2]`.

## v0.5.1 (2020-05-15)

### Bug Fixes

* Minor reorganisation of the bootstraping code.

## v0.5.0 (2019-12-08)

### Breaking Changes

* Support for Julia 0.7 was dropped, now the minimum version required is Julia
  v1.0.

## v0.4.0 (2018-08-23)

### Breaking Changes

* Support for Julia 0.6 was dropped.

## v0.3.1 (2018-07-28)

* Minor performance improvements
* New documentation at https://juliaastro.github.io/LombScargle.jl/stable/

## v0.3.0 (2017-04-29)

### New Features

* You can pre-plan a periodogram before actually performing it using
  `LombScargle.plan` function.  This computes some quantities needed afterwards
  and pre-allocate the memory for the actual computation of the periodogram.
  The speed-up is particularly relevant for the fast method.
* Add `flags` and `timelimit` optional keywords to `lombscargle` function, to
  set the FFTW planner flags and the time limit.
* Package license changed to BSD 3-clause "New" or "Revised".

### Breaking Changes

* Support for Julia 0.4 and 0.5 was dropped.
* The `normalization` keyword of `lombscargle` function now must be a `Symbol`
  (instead of `AbstractString`), the default being `:standard` (instead of
  `"standard"`).  The same normalizations as before are supported, the names
  kept the same capitalization.

### Improvements

This version faced several performance improvements, in particular to
`lombscargle` and `LombScargle.model` functions, besides the pre-planning
feature.

* The fast method of `lombscarlge` now is faster, the larger the size of input
  array, the larger the improvement.  In addition, the fast Fourier transform
  computed internally with FFTW library can take advantage of multi-threading
  (call `FFTW.set_num_threads(n)` to use `n` threads) in order to speed-up
  computation.  However, please note that the running time will not scale as `n`
  because computation of the FFT is only a part of the algorithm.  The memory
  footprint of this function is also considerably lower.  To give you an idea of
  the improvement, for an input of 100000 datapoints, a pre-planned periodogram
  is 70% faster than a (non-planned) periodogram in previous version and
  requires almost 90% less memory.  With 4 FFTW threads the speed-up is of 80%.
* The two non-fast methods now are about 20%-30% faster, thanks to the use of
  `sincos` function from math library.  These methods now also support Julia’s
  native
  [multi-threading](http://docs.julialang.org/en/stable/manual/parallel-computing/#multi-threading-experimental).
  Run Julia with `n` threads (e.g., `JULIA_NUM_THREADS=4 julia` for 4 threads)
  in order to gain an `n`-fold scaling.  These functions also eat considerably
  less memory: if the periodogram is pre-planned, all operations are then
  performed in-place, so memory usage of the periodogram only is independent of
  input size.
* The `LombScargle.model` function is now a bit faster and less memory-greedy
  than before.

### Bug Fixes

* PSD normalization with heteroskedastic errors has been fixed.

## v0.2.0 (2016-12-07)

### Breaking Changes

* The fast method is now used when the frequency grid is evenly spaced (a
  `Range` object), no matter what the `times` vector is.  The previous behavior
  was due to a wrong interpretation of the applicability of the method.

### Bug Fixes

* `Periodogram` type now has 4 parameters, one for the type of each field.  Now
  `power`, `freq`, and `times` fields need not to have all the same floating
  point type.
* In the non-fast variant of the Generalised Lomb–Scargle method, when
  `fit_mean` and/or `center_data` are `true`, pre-center the data by subtracting
  from the signal the weighted average of the signal itself, instead of the
  arithmetic mean.

## v0.1.2 (2016-10-17)

### New Features

* New function for performing bootstrap resampling: `LombScargle.bootstrap`.
  The `fap` and `fapinv` functions have now new methods to estimate false-alarm
  probability and its inverse from a bootstrap sample.
* New utilities: `period`, `periodpower`, `findmaxperiod`.

### Bug Fixes

* Fix power in the standard (i.e., `fast = false` variant) generalised
  Lomb–Scargle algorithm with `fit_mean = true`.  You will find different
  results than before, but for the better, previous results were slightly wrong.

## v0.1.1 (2016-08-20)

### New Features

* New function: `LombScargl.model`.  It gives the best fitting Lomb–Scargle
  model at a given frequency.
* `findmaxfreq` function now can take an optional argument (`interval`) to limit
  the search for the maximum frequency to a certain frequency range.

## v0.1.0 (2016-08-18)

### New Features

* The fast, but approximate, method by Press & Rybicki (1989) has been

src/gls.jl
### gls.jl ---  Perform Generalised Lomb–Scargle periodogram
#
# Copyright (C) 2017 Mosè Giordano.
#
# Maintainer: Mosè Giordano <mose AT gnu DOT org>
# Keywords: periodogram, lomb scargle
#
# This file is a part of LombScargle.jl.
#
# License is BSD 3-clause "New" or "Revised".
#
### Code:

struct GLSPlan{T,A,B<:AbstractVector{T},C,D,E,F,G} <: PeriodogramPlan
    times::A
    signal::B
    freq::C
    sumw::D
    w::E
    y::B
    YY::T
    noise::F
    norm::Symbol
    P::G
end

struct GLSPlan_fit_mean{T,A,B<:AbstractVector{T},C,D,E,F,G,H} <: PeriodogramPlan
    times::A
    signal::B
    freq::C
    sumw::D
    w::E
    y::B
    Y::F
    YY::T
    noise::G
    norm::Symbol
    P::H
end

# Generalised Lomb–Scargle algorithm: this takes into account uncertainties and
# fit the mean of the signal.  This is implemented following the recipe by
# * Zechmeister, M., Kürster, M. 2009, A&A, 496, 577  (URL:
#   http://dx.doi.org/10.1051/0004-6361:200811296,
#   Bibcode: http://adsabs.harvard.edu/abs/2009A%26A...496..577Z)
# In addition, some tricks suggested by
# * Press, W. H., Rybicki, G. B. 1989, ApJ, 338, 277 (URL:
#   http://dx.doi.org/10.1086/167197,
#   Bibcode: http://adsabs.harvard.edu/abs/1989ApJ...338..277P)
# to make computation faster are adopted.

function _generalised_lombscargle!(P, freqs, times, y, w, Y, YY, nil)
    @inbounds Threads.@threads for n in eachindex(freqs)
        ω = freqs[n] * 2 * pi
        # Find τ for current angular frequency
        C = S = CS = CC = nil
        @inbounds for i in eachindex(times)
            W    = w[i]
            s, c = sincos(ω * times[i])
            CS  += W*c*s
            CC  += W*c*c
            C   += W*c
            S   += W*s
        end
        CS -= C*S
        SS  = 1 - CC - S*S
        CC -= C*C
        ωτ   = atan(2CS, CC - SS) / 2
        # Now we can compute the power
        YC_τ = YS_τ = CC_τ = nil
        @inbounds for i in eachindex(times)
            W     = w[i]
            s, c  = sincos(ω*times[i] - ωτ)
            YC_τ += W*y[i]*c
            YS_τ += W*y[i]*s
            CC_τ += W*c*c
        end
        # "C_τ" and "S_τ" are computed following equation (7) of Press &
        # Rybicki, this formula simply comes from angle difference trigonometric
        # identities.
        sin_ωτ, cos_ωτ = sincos(ωτ)
        C_τ    = C*cos_ωτ + S*sin_ωτ
        S_τ    = S*cos_ωτ - C*sin_ωτ
        YC_τ  -= Y*C_τ
        YS_τ  -= Y*S_τ
        SS_τ   = 1 - CC_τ - S_τ*S_τ
        CC_τ  -= C_τ*C_τ
        P[n] = (abs2(YC_τ)/CC_τ + abs2(YS_τ)/SS_τ)/YY
    end
    return P
end

function _generalised_lombscargle!(P, freqs, times, y, w, YY, nil)
    @inbounds Threads.@threads for n in eachindex(freqs)
        ω = freqs[n] * 2 * pi
        # Find τ for current angular frequency
        C = S = CS = CC = nil
        @inbounds for i in eachindex(times)
            W    = w[i]
            s, c = sincos(ω * times[i])
            CS  += W*c*s
            CC  += W*c*c
        end
        ωτ   = atan(2CS, 2CC - 1) / 2
        # Now we can compute the power
        YC_τ = YS_τ = CC_τ = nil
        @inbounds for i in eachindex(times)
            W     = w[i]
            s, c  = sincos(ω*times[i] - ωτ)
            YC_τ += W*y[i]*c
            YS_τ += W*y[i]*s
            CC_τ += W*c*c
        end
        # P[n] should be (abs2(YC_τ)/CC_τ + abs2(YS_τ)/SS_τ)/YY, but we guard the values of
        # CC_τ and SS_τ in case they're zeros (note that SS_τ = 1 - CC_τ).
        frac_C = ifelse(iszero(CC_τ), nil, abs2(YC_τ) / CC_τ)
        frac_S = ifelse(CC_τ == 1,    nil, abs2(YS_τ) / (1 - CC_τ))
        P[n] = (frac_C + frac_S)/YY
    end
    return P
end

_periodogram!(P::AbstractVector, times, p::GLSPlan_fit_mean) =
    _generalised_lombscargle!(P, p.freq, times, p.y, p.w, p.Y, p.YY, zero(eltype(p.P)))
_periodogram!(P::AbstractVector, times, p::GLSPlan) =
    _generalised_lombscargle!(P, p.freq, times, p.y, p.w, p.YY, zero(eltype(p.P)))
